File: lib/geo.js
// UTM <-> WGS84 (Transverse Mercator, WGS84 ellipsoid)
// Zero dependencies. Works in both main thread and Web Workers.

const a   = 6378137.0;
const f   = 1 / 298.257223563;
const e2  = 2 * f - f * f;
const ep2 = e2 / (1 - e2);
const k0  = 0.9996;
const e1  = (1 - Math.sqrt(1 - e2)) / (1 + Math.sqrt(1 - e2));

const M1 = 1 - e2/4 - 3*e2*e2/64 - 5*e2*e2*e2/256;
const M2 = 3*e2/8 + 3*e2*e2/32 + 45*e2*e2*e2/1024;
const M3 = 15*e2*e2/256 + 45*e2*e2*e2/1024;
const M4 = 35*e2*e2*e2/3072;

function centralMeridianRad(zone) {
    return ((zone - 1) * 6 - 180 + 3) * Math.PI / 180;
}

function meridionalArc(phi) {
    return a * (M1*phi - M2*Math.sin(2*phi) + M3*Math.sin(4*phi) - M4*Math.sin(6*phi));
}

// [lon, lat] degrees -> [easting, northing]
export function wgs84ToUtm(lon, lat, zone, isNorth) {
    const rad = Math.PI / 180;
    const phi = lat * rad;
    const lambda0 = centralMeridianRad(zone);
    const dlambda = lon * rad - lambda0;

    const sinPhi = Math.sin(phi), cosPhi = Math.cos(phi), tanPhi = Math.tan(phi);
    const N = a / Math.sqrt(1 - e2 * sinPhi * sinPhi);
    const T = tanPhi * tanPhi;
    const C = ep2 * cosPhi * cosPhi;
    const A = cosPhi * dlambda;
    const M = meridionalArc(phi);

    const A2 = A*A, A3 = A2*A, A4 = A2*A2, A5 = A4*A, A6 = A4*A2;

    const easting = k0 * N * (
        A + (1 - T + C) * A3/6
        + (5 - 18*T + T*T + 72*C - 58*ep2) * A5/120
    ) + 500000;

    let northing = k0 * (
        M + N * tanPhi * (
            A2/2
            + (5 - T + 9*C + 4*C*C) * A4/24
            + (61 - 58*T + T*T + 600*C - 330*ep2) * A6/720
        )
    );

    if (!isNorth) northing += 10000000;

    return [easting, northing];
}

// [easting, northing] -> [lon, lat] degrees
export function utmToWgs84(easting, northing, zone, isNorth) {
    const x = easting - 500000;
    const y = isNorth ? northing : northing - 10000000;

    const M = y / k0;
    const mu = M / (a * M1);

    const phi1 = mu
        + (3*e1/2 - 27*e1*e1*e1/32) * Math.sin(2*mu)
        + (21*e1*e1/16 - 55*e1*e1*e1*e1/32) * Math.sin(4*mu)
        + (151*e1*e1*e1/96) * Math.sin(6*mu)
        + (1097*e1*e1*e1*e1/512) * Math.sin(8*mu);

    const sinPhi1 = Math.sin(phi1), cosPhi1 = Math.cos(phi1), tanPhi1 = Math.tan(phi1);
    const es = e2 * sinPhi1 * sinPhi1;
    const N1 = a / Math.sqrt(1 - es);
    const T1 = tanPhi1 * tanPhi1;
    const C1 = ep2 * cosPhi1 * cosPhi1;
    const R1 = a * (1 - e2) / Math.pow(1 - es, 1.5);
    const D = x / (N1 * k0);

    const D2 = D*D, D3 = D2*D, D4 = D2*D2, D5 = D4*D, D6 = D4*D2;

    const lat = phi1 - (N1 * tanPhi1 / R1) * (
        D2/2
        - (5 + 3*T1 + 10*C1 - 4*C1*C1 - 9*ep2) * D4/24
        + (61 + 90*T1 + 298*C1 + 45*T1*T1 - 252*ep2 - 3*C1*C1) * D6/720
    );

    const lon = (
        D - (1 + 2*T1 + C1) * D3/6
        + (5 - 2*C1 + 28*T1 - 3*C1*C1 + 8*ep2 + 24*T1*T1) * D5/120
    ) / cosPhi1;

    const lambda0 = centralMeridianRad(zone);

    return [
        (lambda0 + lon) * 180 / Math.PI,
        lat * 180 / Math.PI
    ];
}

export function utmParams(epsg) {
    return { zone: epsg % 100, isNorth: epsg < 32700 };
}

export function metersToDegreesLat(m) {
    return m / 110540;
}

export function metersToDegreesLon(m, lat) {
    return m / (111320 * Math.cos(lat * Math.PI / 180));
}

File: core/src/geo.rs
//! utm <-> wgs84 (transverse mercator, wgs84 ellipsoid) + mgrs/epsg helpers.
//! 1:1 port of lib/geo.js — zero deps, identical formulae.

const A: f64 = 6378137.0;
const F: f64 = 1.0 / 298.257223563;
const E2: f64 = 2.0 * F - F * F;
const K0: f64 = 0.9996;

fn ep2() -> f64 {
    E2 / (1.0 - E2)
}
fn e1() -> f64 {
    (1.0 - (1.0 - E2).sqrt()) / (1.0 + (1.0 - E2).sqrt())
}

fn m1() -> f64 {
    1.0 - E2 / 4.0 - 3.0 * E2 * E2 / 64.0 - 5.0 * E2 * E2 * E2 / 256.0
}
fn m2() -> f64 {
    3.0 * E2 / 8.0 + 3.0 * E2 * E2 / 32.0 + 45.0 * E2 * E2 * E2 / 1024.0
}
fn m3() -> f64 {
    15.0 * E2 * E2 / 256.0 + 45.0 * E2 * E2 * E2 / 1024.0
}
fn m4() -> f64 {
    35.0 * E2 * E2 * E2 / 3072.0
}

fn central_meridian_rad(zone: i32) -> f64 {
    ((zone - 1) as f64 * 6.0 - 180.0 + 3.0) * core::f64::consts::PI / 180.0
}

fn meridional_arc(phi: f64) -> f64 {
    A * (m1() * phi - m2() * (2.0 * phi).sin() + m3() * (4.0 * phi).sin()
        - m4() * (6.0 * phi).sin())
}

/// [lon, lat] degrees -> [easting, northing]
pub fn wgs84_to_utm(lon: f64, lat: f64, zone: i32, is_north: bool) -> (f64, f64) {
    let rad = core::f64::consts::PI / 180.0;
    let phi = lat * rad;
    let lambda0 = central_meridian_rad(zone);
    let dlambda = lon * rad - lambda0;

    let (sin_phi, cos_phi, tan_phi) = (phi.sin(), phi.cos(), phi.tan());
    let n = A / (1.0 - E2 * sin_phi * sin_phi).sqrt();
    let t = tan_phi * tan_phi;
    let c = ep2() * cos_phi * cos_phi;
    let aa = cos_phi * dlambda;
    let m = meridional_arc(phi);

    let (a2, a3, a4, a5, a6) = (aa * aa, aa * aa * aa, aa.powi(4), aa.powi(5), aa.powi(6));

    let easting = K0
        * n
        * (aa
            + (1.0 - t + c) * a3 / 6.0
            + (5.0 - 18.0 * t + t * t + 72.0 * c - 58.0 * ep2()) * a5 / 120.0)
        + 500000.0;

    let mut northing = K0
        * (m + n
            * tan_phi
            * (a2 / 2.0
                + (5.0 - t + 9.0 * c + 4.0 * c * c) * a4 / 24.0
                + (61.0 - 58.0 * t + t * t + 600.0 * c - 330.0 * ep2()) * a6 / 720.0));

    if !is_north {
        northing += 10000000.0;
    }
    (easting, northing)
}

/// [easting, northing] -> [lon, lat] degrees
pub fn utm_to_wgs84(easting: f64, northing: f64, zone: i32, is_north: bool) -> (f64, f64) {
    let x = easting - 500000.0;
    let y = if is_north {
        northing
    } else {
        northing - 10000000.0
    };

    let m = y / K0;
    let mu = m / (A * m1());
    let (e1, ep2) = (e1(), ep2());

    let phi1 = mu
        + (3.0 * e1 / 2.0 - 27.0 * e1.powi(3) / 32.0) * (2.0 * mu).sin()
        + (21.0 * e1 * e1 / 16.0 - 55.0 * e1.powi(4) / 32.0) * (4.0 * mu).sin()
        + (151.0 * e1.powi(3) / 96.0) * (6.0 * mu).sin()
        + (1097.0 * e1.powi(4) / 512.0) * (8.0 * mu).sin();

    let (sin1, cos1, tan1) = (phi1.sin(), phi1.cos(), phi1.tan());
    let es = E2 * sin1 * sin1;
    let n1 = A / (1.0 - es).sqrt();
    let t1 = tan1 * tan1;
    let c1 = ep2 * cos1 * cos1;
    let r1 = A * (1.0 - E2) / (1.0 - es).powf(1.5);
    let d = x / (n1 * K0);

    let (d2, d3, d4, d5, d6) = (d * d, d.powi(3), d.powi(4), d.powi(5), d.powi(6));

    let lat = phi1
        - (n1 * tan1 / r1)
            * (d2 / 2.0 - (5.0 + 3.0 * t1 + 10.0 * c1 - 4.0 * c1 * c1 - 9.0 * ep2) * d4 / 24.0
                + (61.0 + 90.0 * t1 + 298.0 * c1 + 45.0 * t1 * t1 - 252.0 * ep2 - 3.0 * c1 * c1)
                    * d6
                    / 720.0);

    let lon = (d - (1.0 + 2.0 * t1 + c1) * d3 / 6.0
        + (5.0 - 2.0 * c1 + 28.0 * t1 - 3.0 * c1 * c1 + 8.0 * ep2 + 24.0 * t1 * t1) * d5 / 120.0)
        / cos1;

    let lambda0 = central_meridian_rad(zone);
    (
        (lambda0 + lon) * 180.0 / core::f64::consts::PI,
        lat * 180.0 / core::f64::consts::PI,
    )
}

/// (zone, is_north) from a utm epsg code.
pub fn utm_params(epsg: i32) -> (i32, bool) {
    (epsg % 100, epsg < 32700)
}

/// utm epsg from an mgrs grid code ('MGRS-39RWN' or '39RWN'). cdse stac omits
/// proj:epsg; the tile encodes it (zone = leading digits, band N–X north / C–M south).
pub fn epsg_from_mgrs(grid: &str) -> i32 {
    let t = grid.strip_prefix("MGRS-").unwrap_or(grid);
    let digits: String = t.chars().take_while(|c| c.is_ascii_digit()).collect();
    let zone: i32 = digits.parse().unwrap_or(0);
    let band = t.as_bytes().get(digits.len()).copied().unwrap_or(b'N');
    (if band >= b'N' { 32600 } else { 32700 }) + zone
}

pub fn meters_to_degrees_lat(m: f64) -> f64 {
    m / 110540.0
}
pub fn meters_to_degrees_lon(m: f64, lat: f64) -> f64 {
    m / (111320.0 * (lat * core::f64::consts::PI / 180.0).cos())
}

/// expand a bbox [w,s,e,n] by `km` on every side (latitude-corrected).
pub fn pad_bbox(b: [f64; 4], km: f64) -> [f64; 4] {
    if km == 0.0 {
        return b;
    }
    let [w, s, e, n] = b;
    let dlat = km / 111.0;
    let dlon = km / (111.0 * ((s + n) / 2.0 * core::f64::consts::PI / 180.0).cos());
    [w - dlon, s - dlat, e + dlon, n + dlat]
}

/// approximate bbox area in km² (latitude-corrected) — the web api's abuse cap.
pub fn bbox_area_km2(b: [f64; 4]) -> f64 {
    let [w, s, e, n] = b;
    (n - s) * 111.0 * (e - w) * 111.0 * ((s + n) / 2.0 * core::f64::consts::PI / 180.0).cos()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn roundtrip_utm() {
        let (lon, lat) = (-102.0208, 32.0071);
        let (e, n) = wgs84_to_utm(lon, lat, 13, true);
        let (lon2, lat2) = utm_to_wgs84(e, n, 13, true);
        assert!((lon - lon2).abs() < 1e-7 && (lat - lat2).abs() < 1e-7);
    }
    #[test]
    fn mgrs_epsg() {
        assert_eq!(epsg_from_mgrs("MGRS-39RWN"), 32639);
        assert_eq!(epsg_from_mgrs("13SDV"), 32613);
        assert_eq!(epsg_from_mgrs("39CWN"), 32739); // C band → south
    }
    #[test]
    fn params() {
        assert_eq!(utm_params(32613), (13, true));
        assert_eq!(utm_params(32739), (39, false));
    }
}
